### PR TITLE
[AFID-348] Execution context and model refactor

### PR DESCRIPTION
**Pull request by da-hmrc**

### CODE REVIEW

File: app/uk/gov/hmrc/taxhistory/services/TaxHistoryMongoCacheService.scala [hunk 1]
```diff
@@ -28,50 +28,56 @@ import uk.gov.hmrc.taxhistory.model.api.PayAsYouEarn
...
 
   implicit val mongo: () => DB = mongoDbConnection.db
 
   val cacheRepository = new CacheMongoRepository(mongoSource, expireAfterSeconds, Cache.mongoFormats)
 
-  def createOrUpdate(id: String, key: String, toCache: JsValue): Future[Option[JsValue]] = {
-    cacheRepository.createOrUpdate(Id(id), key, toCache).map(x => x.updateType.savedValue.data.map(_ \ key).map(_.get))
+  def insertOrUpdate(key: (Nino, TaxYear), value: PayAsYouEarn): Future[Option[PayAsYouEarn]] = {
+    val (nino, taxYear) = key
+    val mongoId = nino.nino
```
Comment (Smudgey) [inline app/uk/gov/hmrc/taxhistory/services/TaxHistoryMongoCacheService.scala]: should this be nino.value ?

File: app/uk/gov/hmrc/taxhistory/services/TaxHistoryMongoCacheService.scala [hunk 2]
```diff
@@ -28,50 +28,56 @@ import uk.gov.hmrc.taxhistory.model.api.PayAsYouEarn
...
+    val mongoKey = taxYear.currentYear.toString
+    cacheRepository
+      .createOrUpdate(Id(mongoId), mongoKey, Json.toJson[PayAsYouEarn](value))
+      .map(_.updateType.savedValue.data.map(_ \ mongoKey).map(_.get.as[PayAsYouEarn]))
   }
 
-  def findById(id: String, taxYear: Int): Future[Option[JsValue]] = {
-    cacheRepository.findById(Id(id)).map(_.flatMap(_.data).map(_ \ taxYear.toString).flatMap(_.toOption))
+  private def getFromRepository(nino: Nino, taxYear: TaxYear): Future[Option[PayAsYouEarn]] = {
+    val mongoId = nino.nino
```
Comment (Smudgey) [inline app/uk/gov/hmrc/taxhistory/services/TaxHistoryMongoCacheService.scala]: Same as above

File: test/uk/gov/hmrc/taxhistory/services/helper/EmploymentHistoryServiceHelperSpec.scala [hunk 3]
```diff
@@ -113,121 +115,103 @@ class EmploymentHistoryServiceHelperSpec extends PlaySpec with MockitoSugar with
...
-      val benefits1 = merged.benefits.get(employment1.employmentId.toString)
+      merged.benefits.size mustBe 1
+      val Some(benefits1) = merged.benefits.get(employment1.employmentId.toString)
       benefits1.size mustBe 1
       benefits1 must contain(companyBenefit)
 
-      merged.payAndTax.get.size mustBe 1
-      val payAndTax1 = merged.payAndTax.get(employment1.employmentId.toString)
+      merged.payAndTax.size mustBe 1
+      val Some(payAndTax1) = merged.payAndTax.get(employment1.employmentId.toString)
```
Comment (Smudgey) [inline test/uk/gov/hmrc/taxhistory/services/helper/EmploymentHistoryServiceHelperSpec.scala]: I've not seen this syntax when declaring a val before, can you explain it?

Specifically, it's this part: val Some(payAndTax1) = ...

What happens when the value assigned to it is None?

Comment (da-hmrc) [reply test/uk/gov/hmrc/taxhistory/services/helper/EmploymentHistoryServiceHelperSpec.scala]: It's very much like pattern matching. When the match fails (in this case, when it's None), an exception is thrown (same as before the changes, when there was an `Option.get`)
It doesn't make it safe, just arguably clearer (and it was the quicker way to change the test so it compiled again). But this is a test, in production we shouldn't do it, we should check for None.

File: app/uk/gov/hmrc/taxhistory/services/TaxHistoryMongoCacheService.scala (hunk 1 above)
Comment (da-hmrc) [reply app/uk/gov/hmrc/taxhistory/services/TaxHistoryMongoCacheService.scala]: The two are the same (https://github.com/hmrc/domain/blob/eecf5a5169a9f2205cc89def487e0b9121fd5af8/src/main/scala/uk/gov/hmrc/domain/Nino.scala#L27)
I'm not sure which is the 'correct' way of referencing it!

Comment (Smudgey) [reply app/uk/gov/hmrc/taxhistory/services/TaxHistoryMongoCacheService.scala]: Okay, guess it comes down to which is clearer? Do you prefer nino.nino or nino.value ?

Comment (da-hmrc) [reply app/uk/gov/hmrc/taxhistory/services/TaxHistoryMongoCacheService.scala]: Maybe `nino.value` is better.

Comment (Smudgey) [reply app/uk/gov/hmrc/taxhistory/services/TaxHistoryMongoCacheService.scala]: Okay, go for that then

**Pull request merged**